File: stock-research-agent-api/Services/ResearchEngine/DailyReportService.cs
using StockResearchAgent.Api.Models;
using static StockResearchAgent.Api.Services.ResearchEngine.OutcomeEvaluator;

namespace StockResearchAgent.Api.Services.ResearchEngine;

/// <summary>
/// Generates human-readable text reports for morning scans and EOD reviews.
/// </summary>
public class DailyReportService
{
    // -----------------------------------------------------------------------
    // Morning Report
    // -----------------------------------------------------------------------

    public string GenerateMorningReport(List<PredictionCandidate> predictions, List<MarketSnapshot> snapshots)
    {
        var parts = new List<string>();
        var now = DateTime.UtcNow;
        parts.Add($"Morning Research Scan - {now:dddd, MMMM d, yyyy}");
        parts.Add("");

        // Market overview
        var spy = snapshots.FirstOrDefault(s => s.Ticker == "SPY");
        var qqq = snapshots.FirstOrDefault(s => s.Ticker == "QQQ");
        if (spy?.Quote is not null || qqq?.Quote is not null)
        {
            parts.Add("MARKET OVERVIEW");
            if (spy?.Quote is not null)
                parts.Add($"  SPY: ${spy.Quote.Price:F2} ({(spy.Quote.ChangePercent > 0 ? "+" : "")}{spy.Quote.ChangePercent:F2}%)");
            if (qqq?.Quote is not null)
                parts.Add($"  QQQ: ${qqq.Quote.Price:F2} ({(qqq.Quote.ChangePercent > 0 ? "+" : "")}{qqq.Quote.ChangePercent:F2}%)");
            parts.Add("");
        }

        // Data warnings
        var unavailable = snapshots.Where(s => !s.DataAvailability.MarketDataAvailable).Select(s => s.Ticker).ToList();
        if (unavailable.Count > 0)
        {
            parts.Add($"DATA WARNINGS: Market data unavailable for {string.Join(", ", unavailable)}");
            parts.Add("");
        }

        // Predictions summary
        var bullish = predictions.Count(p => p.PredictionType == PredictionType.bullish);
        var bearish = predictions.Count(p => p.PredictionType == PredictionType.bearish);
        var neutral = predictions.Count(p => p.PredictionType == PredictionType.neutral);
        var watchOnly = predictions.Count(p => p.PredictionType == PredictionType.watch_only);

        parts.Add($"PREDICTIONS GENERATED: {predictions.Count} total");
        parts.Add($"  Bullish: {bullish} | Bearish: {bearish} | Neutral: {neutral} | Watch-only: {watchOnly}");
        parts.Add("");

        // Top predictions
        var topPicks = predictions.Where(p => p.ConfidenceScore >= 20)
            .OrderByDescending(p => p.ConfidenceScore).Take(5).ToList();
        if (topPicks.Count > 0)
        {
            parts.Add("TOP PREDICTIONS (by confidence):");
            foreach (var p in topPicks)
            {
                parts.Add($"  {p.Ticker} - {p.PredictionType.ToString().ToUpper()} (conf: {p.ConfidenceScore}, risk: {p.RiskScore})");
                if (p.EntryReferencePrice is not null) parts.Add($"    Entry ref: ${p.EntryReferencePrice:F2}");
                parts.Add($"    Reason: {(p.PredictionReason.Length > 150 ? p.PredictionReason[..150] : p.PredictionReason)}");
                if (p.MissingDataWarnings.Count > 0) parts.Add($"    Missing: {string.Join("; ", p.MissingDataWarnings)}");
            }
            parts.Add("");
        }

        // High-impact catalysts
        var allNews = snapshots.SelectMany(s => s.NewsContext.Select(n => (n, s.Ticker))).ToList();
        var highImpact = allNews.Where(x => x.n.ImportanceScore >= 7).Take(5).ToList();
        if (highImpact.Count > 0)
        {
            parts.Add("HIGH-IMPACT CATALYSTS:");
            foreach (var (n, ticker) in highImpact)
                parts.Add($"  [{ticker}] {n.Title} ({n.CatalystType ?? "news"}, imp={n.ImportanceScore})");
            parts.Add("");
        }

        parts.Add("This is automated research, not financial advice. All predictions are watchlist candidates only.");
        return string.Join("\n", parts);
    }

    // -----------------------------------------------------------------------
    // End-of-Day Report
    // -----------------------------------------------------------------------

    public string GenerateEndOfDayReport(List<EvaluationResult> evaluated, List<string> skipped)
    {
        var parts = new List<string>();
        var now = DateTime.UtcNow;
        parts.Add($"End-of-Day Review - {now:dddd, MMMM d, yyyy}");
        parts.Add("");
        parts.Add($"EVALUATIONS: {evaluated.Count} predictions scored, {skipped.Count} skipped");
        parts.Add("");

        if (evaluated.Count > 0)
        {
            var correct = evaluated.Count(e => e.Outcome.DirectionCorrect == true);
            var wrong = evaluated.Count(e => e.Outcome.DirectionCorrect == false);
            var neutralCount = evaluated.Count(e => e.Outcome.DirectionCorrect is null);
            var avgScore = evaluated.Average(e => e.Outcome.OutcomeScore ?? 50);

            parts.Add($"RESULTS: {correct} correct, {wrong} wrong, {neutralCount} neutral/N-A");
            parts.Add($"Average outcome score: {avgScore:F1}/100");
            if (correct + wrong > 0)
                parts.Add($"Direction accuracy: {(double)correct / (correct + wrong) * 100:F1}%");
            parts.Add("");

            parts.Add("DETAILS:");
            foreach (var e in evaluated)
            {
                var o = e.Outcome;
                var moveStr = o.PercentMove is not null ? $"{(o.PercentMove > 0 ? "+" : "")}{o.PercentMove:F2}%" : "N/A";
                var dirStr = o.DirectionCorrect == true ? "CORRECT" : o.DirectionCorrect == false ? "WRONG" : "N/A";
                parts.Add($"  {e.Ticker}: {moveStr} ({dirStr}, score: {o.OutcomeScore?.ToString("F0") ?? "N/A"})");
                if (o.Lesson is not null)
                    parts.Add($"    Lesson: {(o.Lesson.Length > 120 ? o.Lesson[..120] : o.Lesson)}");
            }
            parts.Add("");
        }

        if (skipped.Count > 0)
        {
            parts.Add("SKIPPED:");
            foreach (var s in skipped.Take(10)) parts.Add($"  {s}");
            parts.Add("");
        }

        parts.Add("This is automated research review, not financial advice.");
        return string.Join("\n", parts);
    }
}

File: stock-research-agent/services/researchEngine/dailyReportService.ts
/**
 * Generates human-readable reports for morning scans and EOD reviews.
 * Uses the .NET AI API when available, falls back to rule-based
 * summaries otherwise.
 */

import 'server-only';
import type {
  PredictionCandidateInput,
  MarketSnapshot,
} from './researchEngine.types';

// ---------------------------------------------------------------------------
// Morning Report
// ---------------------------------------------------------------------------

export async function generateMorningReport(
  predictions: PredictionCandidateInput[],
  snapshots: MarketSnapshot[],
): Promise<string> {
  const parts: string[] = [];
  const now = new Date();
  parts.push(`Morning Research Scan - ${now.toLocaleDateString('en-US', { weekday: 'long', month: 'long', day: 'numeric', year: 'numeric' })}`);
  parts.push('');

  // Market overview
  const spySnapshot = snapshots.find((s) => s.ticker === 'SPY');
  const qqqSnapshot = snapshots.find((s) => s.ticker === 'QQQ');
  if (spySnapshot?.quote || qqqSnapshot?.quote) {
    parts.push('MARKET OVERVIEW');
    if (spySnapshot?.quote) {
      parts.push(`  SPY: $${spySnapshot.quote.price.toFixed(2)} (${spySnapshot.quote.changePercent > 0 ? '+' : ''}${spySnapshot.quote.changePercent.toFixed(2)}%)`);
    }
    if (qqqSnapshot?.quote) {
      parts.push(`  QQQ: $${qqqSnapshot.quote.price.toFixed(2)} (${qqqSnapshot.quote.changePercent > 0 ? '+' : ''}${qqqSnapshot.quote.changePercent.toFixed(2)}%)`);
    }
    parts.push('');
  }

  // Data availability
  const unavailable = snapshots.filter((s) => !s.dataAvailability.marketDataAvailable);
  if (unavailable.length > 0) {
    parts.push(`DATA WARNINGS: Market data unavailable for ${unavailable.map((s) => s.ticker).join(', ')}`);
    parts.push('');
  }

  // Predictions summary
  const bullish = predictions.filter((p) => p.predictionType === 'bullish');
  const bearish = predictions.filter((p) => p.predictionType === 'bearish');
  const neutral = predictions.filter((p) => p.predictionType === 'neutral');
  const watchOnly = predictions.filter((p) => p.predictionType === 'watch_only');

  parts.push(`PREDICTIONS GENERATED: ${predictions.length} total`);
  parts.push(`  Bullish: ${bullish.length} | Bearish: ${bearish.length} | Neutral: ${neutral.length} | Watch-only: ${watchOnly.length}`);
  parts.push('');

  // Top predictions by confidence
  const sorted = [...predictions].sort((a, b) => b.confidenceScore - a.confidenceScore);
  const topPicks = sorted.filter((p) => p.confidenceScore >= 20).slice(0, 5);

  if (topPicks.length > 0) {
    parts.push('TOP PREDICTIONS (by confidence):');
    for (const p of topPicks) {
      parts.push(`  ${p.ticker} - ${p.predictionType.toUpperCase()} (conf: ${p.confidenceScore}, risk: ${p.riskScore})`);
      if (p.entryReferencePrice) parts.push(`    Entry ref: $${p.entryReferencePrice.toFixed(2)}`);
      parts.push(`    Reason: ${p.predictionReason.slice(0, 150)}`);
      if (p.missingDataWarnings.length > 0) {
        parts.push(`    Missing: ${p.missingDataWarnings.join('; ')}`);
      }
    }
    parts.push('');
  }

  // Catalysts
  const allNews = snapshots.flatMap((s) => s.newsContext.map((n) => ({ ...n, ticker: s.ticker })));
  const highImpact = allNews.filter((n) => n.importanceScore >= 7).slice(0, 5);
  if (highImpact.length > 0) {
    parts.push('HIGH-IMPACT CATALYSTS:');
    for (const n of highImpact) {
      parts.push(`  [${n.ticker}] ${n.title} (${n.catalystType ?? 'news'}, imp=${n.importanceScore})`);
    }
    parts.push('');
  }

  parts.push('This is automated research, not financial advice. All predictions are watchlist candidates only.');

  return parts.join('\n');
}

// ---------------------------------------------------------------------------
// End-of-Day Report
// ---------------------------------------------------------------------------

interface EvalResult {
  predictionId: string;
  ticker: string;
  outcome: {
    startPrice: number | null;
    closePrice: number | null;
    percentMove: number | null;
    directionCorrect: boolean | null;
    outcomeScore: number | null;
    outcomeSummary: string | null;
    lesson: string | null;
  };
}

export async function generateEndOfDayReport(
  evaluated: EvalResult[],
  skipped: string[],
): Promise<string> {
  const parts: string[] = [];
  const now = new Date();
  parts.push(`End-of-Day Review - ${now.toLocaleDateString('en-US', { weekday: 'long', month: 'long', day: 'numeric', year: 'numeric' })}`);
  parts.push('');

  parts.push(`EVALUATIONS: ${evaluated.length} predictions scored, ${skipped.length} skipped`);
  parts.push('');

  if (evaluated.length > 0) {
    const correct = evaluated.filter((e) => e.outcome.directionCorrect === true).length;
    const wrong = evaluated.filter((e) => e.outcome.directionCorrect === false).length;
    const neutral = evaluated.filter((e) => e.outcome.directionCorrect === null).length;
    const avgScore = evaluated.reduce((sum, e) => sum + (e.outcome.outcomeScore ?? 50), 0) / evaluated.length;

    parts.push(`RESULTS: ${correct} correct, ${wrong} wrong, ${neutral} neutral/N-A`);
    parts.push(`Average outcome score: ${avgScore.toFixed(1)}/100`);
    if (correct + wrong > 0) {
      parts.push(`Direction accuracy: ${((correct / (correct + wrong)) * 100).toFixed(1)}%`);
    }
    parts.push('');

    parts.push('DETAILS:');
    for (const e of evaluated) {
      const o = e.outcome;
      const moveStr = o.percentMove !== null ? `${o.percentMove > 0 ? '+' : ''}${o.percentMove.toFixed(2)}%` : 'N/A';
      const dirStr = o.directionCorrect === true ? 'CORRECT' : o.directionCorrect === false ? 'WRONG' : 'N/A';
      parts.push(`  ${e.ticker}: ${moveStr} (${dirStr}, score: ${o.outcomeScore ?? 'N/A'})`);
      if (o.lesson) parts.push(`    Lesson: ${o.lesson.slice(0, 120)}`);
    }
    parts.push('');
  }

  if (skipped.length > 0) {
    parts.push('SKIPPED:');
    for (const s of skipped.slice(0, 10)) {
      parts.push(`  ${s}`);
    }
    parts.push('');
  }

  parts.push('This is automated research review, not financial advice.');
  return parts.join('\n');
}
